### PR TITLE
[MAINTENANCE] Unit tests to ensure coverage over `Datasource` caching in `DataContext`

### PR DESCRIPTION
**Pull request by cdkini**
Changes proposed in this pull request:
- Cover various scenarios involving the `_cached_datasources` cache on the `DataContext`.

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

### CODE REVIEW

**Review by Kilo59 — APPROVED** [review]

File: tests/data_context/test_data_context_datasources.py [hunk 1]
```diff
@@ -54,3 +56,129 @@ def test_data_context_instantiates_inline_store_backend_with_filesystem_config(
...
+
+@pytest.mark.unit
+def test_delete_datasource_updates_cache(
+    in_memory_runtime_context: BaseDataContext,
+) -> None:
+    context = in_memory_runtime_context
+
+    name = context.list_datasources()[0]["name"]
+
+    # If the value is in the cache, no store methods should be invoked
```
Comment (anthonyburdi) [inline tests/data_context/test_data_context_datasources.py:137]: Interesting - is this current behavior? That delete only updates the cache and not the store?

**Review by anthonyburdi — CHANGES_REQUESTED** [review]
Looks good but see my question about datasource delete before merging.

File: tests/data_context/test_data_context_datasources.py (hunk 1 above)
Comment (cdkini) [reply tests/data_context/test_data_context_datasources.py:137]: we need to use `save_changes=True` or work with a `DataContext` if we want to utilize the store. This is current behavior!

Comment (anthonyburdi) [reply tests/data_context/test_data_context_datasources.py:137]: Good call! Mind adding that to a test docstring just to clarify?

**Review by anthonyburdi — APPROVED** [review]

**Pull request merged**